Add server extensions documentation (#771)
Shopify/ruby-lsp: SERVER_EXTENSIONS.md

# Server extensions

> **WARNING**
> The Ruby LSP server extensions system is currently experimental and subject to changes in the API

## Motivation and goals

Editor features that are specific to certain tools or frameworks can be incredibly powerful. Typically, language servers
are aimed at providing features for a particular programming language (like Ruby!) and not specific tools. This is
reasonable since not every programmer uses the same combination of tools.

Including tool specific functionality in the Ruby LSP would not scale well given the large number of tools in the
ecosystem. It would also create a bottleneck for authors to push new features. Building separate tooling, on the other
hand, increases fragmentation which tends to increase the effort required by users to configure their development
environments.

For these reasons, the Ruby LSP ships with a server extension system that authors can use to enhance the behavior of the
base LSP with tool specific functionality, aimed at

> - Allowing gem authors to export Ruby LSP extensions from their own gems

Reviewer: ```suggestion
- Allowing gem authors to extend Ruby LSP's functionality from their own gems
```

Reviewer: Ok, let's leave as-is.

- There are two types of LSP requests: automatic (e.g.: semantic highlighting) and user initiated (go to definition).
The performance of automatic requests is critical for responsiveness as they are executed every time the user types
- Avoid duplicate work where possible. If something can be computed once and memoized, like configurations, do it
- Do not mutate LSP state directly. Extensions sometimes have access to important state such as document objects, which
should never be mutated directly, but instead through the mechanisms provided by the LSP specification - like text edits
- Do not overnotify users. It's generally annoying and diverts attention from the current task

## Building a Ruby LSP extension

**Note**: the Ruby LSP uses Sorbet. We recommend using Sorbet in extensions as well, which allows authors to benefit
from types declared by the Ruby LSP.

As an example, check out [Ruby LSP Rails](https://github.com/Shopify/ruby-lsp-rails), which is a Ruby LSP extension to
provide Rails related features.

### Activating the extension

The Ruby LSP discovers extensions based on the existence of an `extension.rb` file placed inside a `ruby_lsp` folder.

> For example, `my_gem/lib/ruby_lsp/my_gem/extension.rb`. This file must declare the extension class, which can be used to

Reviewer 2: Is this a common pattern to repeat the gem name?

perform any necessary activation when the server starts.

```ruby
# frozen_string_literal: true

require "ruby_lsp/extension"

module RubyLsp
  module MyGem
    class Extension < ::RubyLsp::Extension
      extend T::Sig

      # Performs any activation that needs to happen once when the language server is booted

>       sig { override.void }

Author: Good point. For now, I added a link to the Sorbet website where we recommend it.

      sig { params(emitter: RubyLsp::EventEmitter, message_queue: Thread::Queue).void }
      def initialize(emitter, message_queue)
        super

        @response = T.let(nil, ResponseType)

        # Register that this listener will handle `on_const` events (i.e.: whenever a constant is found in the code)
        emitter.register(self, :on_const)
      end

      # Listeners must define methods for each event they registered with the emitter. In this case, we have to define
      # `on_const` to specify what this listener should do every time we find a constant
      sig { params(node: SyntaxTree::Const).void }
      def on_const(node)
        # Certain helpers are made available to listeners to build LSP responses. The classes under `RubyLsp::Interface`
        # are generally used to build responses and they match exactly what the specification requests.
        contents = RubyLsp::Interface::MarkupContent.new(kind: "markdown", value: "Hello!")
        @response = RubyLsp::Interface::Hover.new(range: range_from_syntax_tree_node(node), contents: contents)
      end
    end
  end
end
```

### Registering formatters

> Gems may also register a formatter to be used by the Ruby LSP. To do that, the extension must create a formatter runner

Reviewer: ```suggestion
Gems may also register a provide to be used by the Ruby LSP. To do that, the extension must create a formatter runner
```
(since we say `register` in the next sentence).

  # IMPORTANT: This method must return the formatted document source without mutating the original one in document
  def run(uri, document)
    source = document.source
    formatted_source = format_the_source_using_my_formatter(source)
    formatted_source
  end
end
```

### Sending notifications to the client

Sometimes, requests may want to send asynchronous information to the client. For example, a slow request may want to
indicate progress. To send notifications, all listeners have access to the message queue, where they can push
notifications to the client.

```ruby
class MyListener < ::RubyLsp::Listener
  def initialize(emitter, message_queue)
    super

    @message_queue << Notification.new(
      message: "$/progress",
      params: Interface::ProgressParams.new(
        token: "progress-token-id",
        value: Interface::WorkDoneProgressBegin.new(kind: "begin", title: "Starting slow work!"),
      )
    )
  end
end
```

### Ensuring consistent documentation

The Ruby LSP exports a rake task to help authors make sure all of their listeners are documented and include demos and

> examples of the feature in action. Configure the rake task and run `bundle exec rake ruby_lsp:check_docs` on CI to

Reviewer: ```suggestion
examples of the feature in action. Configure the Rake task and run `bundle exec rake ruby_lsp:check_docs` on CI to
```

# Server extensions

> The Ruby LSP server extensions system is currently experimental and subject to changes in the API

## Motivation and goals

Having editor features that are specific to certain tools can be incredibly helpful for programmers. For example, Rails

> related features or specific formatters. Typically, language servers are aimed at providing features for a particular

Author: I made this a bit shorter, but I still find value in explaining our motivation. Let me know what you think.

programmer will use the same set of tools.

Trying to include tool specific functionality directly in the Ruby LSP would not scale well given the large number of
tools in the Ruby ecosystem. It would also create a bottleneck for tool authors since all changes would need to be
merged into the Ruby LSP's main codebase.

Creating language servers for each tool is also not an adequate solution (e.g.: Rails LSP, Sinatra LSP). While this
might sometimes be easier for the maintainers, it places the burden on the user to figure out which language servers to
install to get a complete set of features and how to configure them so that there are no conflicts.

For these reasons, the Ruby LSP ships with a server extension system that gem authors can use to enhance the behavior of
the base language server with tool specific functionality. The goals of the extension system are

- To allow gem authors to export Ruby LSP extensions from their own gems
- To allow LSP features to be enhanced by extensions that are present in the application the developer is currently
working on, without requiring extra configuration from the user
- To seamlessly integrate with the base features of the Ruby LSP

> ## Guidelines

Reviewer 3: I personally find the difference between automatic and user-initiated requests important but I wasn't aware of it until reading this. So perhaps adding it as the second point too?

> It's generally annoying and diverts attention from the current task

It may be worth mentioning that we learned this through experiences, not just an opinion.

> ## Guidelines

Author: I added another point related to automatic and user initiated requests.

Regarding the last point, I honestly think it's fine to keep it as is - even if it was an opinion.
